usb: device_next: make HS support compliant with the USB2.0 specification

jfischer-no opened the pull request:
The main goal of this PR is to improve high-speed support and make it USB2.0 specification compliant.
Tested on stm32f723e_disco with few fixes in the driver and board configuration.

jfischer-no committed:
usb: device_next: make HS support compliant with the USB2.0 specification

For specification-compliant high-speed support, we need to support
device quilifiers and other-speed-configuration descriptor requests. We
also need to store different configurations of the class/function
descriptors, which typically only affect the endpoint descriptors. With
this change, the stack expects class/function descriptors to be passed
as an array of struct usb_desc_header pointers to e.g. interface,
interface-specific, and endpoint descriptors, with the last element of
the array pointing to a nil descriptor. And also passed for a specific
speed, for now we support full and high speed configurations.

During instantiation, the class/function implementation must choose the
correct configuration in the full-speed and high-speed descriptor sets
for values such as maximum packet size and bInterval values of interrupt
and isochronous endpoints.

During initialization, the stack reads the highest speed supported by
the controller and uses it to get the appropriate descriptors set
from the instance. If the controller supports only full speed, the stack
configures the class/function descriptor for full speed only, if the
controller supports high speed, the stack configures the descriptors for
high speed only, and a class/function must update the full speed
descriptor during the init callback processing.

During device operation, the class/function implementation must check
the actual speed of the device and use the correct configuration, such
as the endpoint address or maximum packet size.

Signed-off-by: Johann Fischer <johann.fischer@nordicsemi.no>

jfischer-no committed:
usb: device_next: add helpers to get private data and device context

The class implementations should not access the members of the struct
usbd_class_node directly.

Signed-off-by: Johann Fischer <johann.fischer@nordicsemi.no>

jfischer-no committed:
usb: device_next: rework loopback interfaces

This implementation is WIP and has mostly been used to test interface
and endpoint configuration by the stack. For future changes, make the
interfaces more compliant with the USB specification.

Signed-off-by: Johann Fischer <johann.fischer@nordicsemi.no>

jfischer-no committed:
include: usb_ch9: add macros to calculate endpoint bInterval

Add macros to calculate FS/HS endpoint bInterval.

Signed-off-by: Johann Fischer <johann.fischer@nordicsemi.no>

jfischer-no committed:
usb: device_next: add helpers to get highest supported and actual speed

Save actual device speed in the device context and add helpers to get
highest supported and actual speed.

Signed-off-by: Johann Fischer <johann.fischer@nordicsemi.no>

jfischer-no committed:
include: usb: add device qualifier descriptor

Add device qualifier descriptor.

Signed-off-by: Johann Fischer <johann.fischer@nordicsemi.no>

jfischer-no committed:
usb: device_next: introduce and interface to get class/function descriptor

Introduce interface to get class/function descriptor for specific
speed.

Signed-off-by: Johann Fischer <johann.fischer@nordicsemi.no>

tmon-nordic reviewed (changes requested):
Please keep the old mechanism to allow USB Audio 2 for the time being. I will come up with the speed-specific macrobatics in UAC2 later (i.e. after v3.6.0 release) and then the old mechanism (i.e. the `desc` pointer in `struct usbd_class_data`) can be removed (converted classes can simply set the `desc` to NULL).

I don't agree on the nil descriptor at the end of the descriptor sets - it doesn't make sense, just have NULL pointer. 

I believe just running endpoint assignment at initialization for all supported speeds and having separate configuration descriptors for each speed (i.e. the 9 bytes long CONFIGURATION descriptor) would make the code simpler.

tmon-nordic commented on subsys/usb/device_next/class/loopback.c:
```diff
@@ -43,10 +46,133 @@ struct loopback_desc {
+};
 
 #define DEFINE_LOOPBACK_DESCRIPTOR(x, _)					\
 static struct loopback_desc lb_desc_##x = {					\
```
After a second thought, it can be kept as-is now, but in the future separated to allow the descriptors that are constant to reside only in code memory and not in RAM (which means both code and RAM, because initialization data must be in code).

tmon-nordic committed:
usb: device_next: uac2: add descriptor sets macros

Add necessary macros and convert UAC2 descriptor test from descriptor
blob to descriptor set. Currently there is only Full-Speed descriptor
set.

Signed-off-by: Tomasz Moń <tomasz.mon@nordicsemi.no>

tmon-nordic committed:
usb: device_next: uac2: Convert class to descriptor set

Convert to new descriptor interface, present only Full-Speed descriptors
to the USB stack.

Signed-off-by: Tomasz Moń <tomasz.mon@nordicsemi.no>

tmon-nordic committed:
usb: device_next: allow terminating descriptor set with NULL

Do not require nil descriptor at the end of descriptor set because it
serves no other purpose than a sentinel. Just end processing on first
NULL pointer within descriptor set.

Signed-off-by: Tomasz Moń <tomasz.mon@nordicsemi.no>
Signed-off-by: Johann Fischer <johann.fischer@nordicsemi.no>

jfischer-no commented on subsys/usb/device_next/usbd_ch9.c:
```diff
@@ -440,33 +440,85 @@ static int sreq_get_status(struct usbd_contex *const uds_ctx,
-	len = MIN(setup->wLength, net_buf_tailroom(buf));
-	net_buf_add_mem(buf, cfg_desc, MIN(len, cfg_desc->bLength));
+	if (other_cfg) {
+		/*
+		 * Because the structure of the other-speed-configuration is
+		 * the same as a configuration descriptor, and the other speed
+		 * function collection has the same length and the number of
+		 * interfaces and endpoints, we simply copy the configuration
+		 * descriptor and update the type.
+		 * If at some point the number of interfaces or endpoints for
+		 * full and high speed descritpors in a class implementation
+		 * becomes different, we need to revisit this and compute the
+		 * configuration descriptor properties on the fly.
+		 */
```
This should be considered, I wanted it, but unless there is a case where the number of interfaces/endpoints are different, then it is probably just a luxury.

jfischer-no commented on subsys/usb/device_next/usbd_init.c:
```diff
@@ -109,26 +109,48 @@ static int unassign_eps(struct usbd_contex *const uds_ctx,
 	uint8_t tmp_nif;
 	int ret;
 
+	/*
+	 * Read the highest speed supported by the controller and use it to get
+	 * the appropriate function descriptor from the instance. If the
+	 * controller only supports full speed, the code below will configure
+	 * the function descriptor for full speed only, and high speed will
+	 * never be used after the device instance is initialized. If the
+	 * controller supports high speed, the code will only configure the
+	 * function's high speed descriptor, and the function implementation
+	 * must update the full speed descriptor during the init callback
+	 * processing, which is required to properly respond to
+	 * other-speed-configuration descriptor requests and for the unlikely
+	 * case where the high speed controller is connected to a full speed bus.
+	 */
```
related to https://github.com/zephyrproject-rtos/zephyr/pull/67749#discussion_r1497506488

tmon-nordic commented on subsys/usb/device_next/class/bt_hci.c:
```diff
@@ -433,10 +445,31 @@ static int bt_hci_ctd(struct usbd_class_node *const c_nd,
+
+	desc->iad.bFirstInterface = desc->if0.bInterfaceNumber;
+	if (usbd_caps_speed(uds_ctx) == USBD_SPEED_HS) {
+		LOG_INF("FS endpoint descriptor needs to be updated");
```
This is only your opinion, not a requirement anywhere in the USB specification that the HS set requires more resources.

tmon-nordic committed:
usb: device_next: Handle class requests the same regardless of debug

When device is not configured, any class or endpoint request must result
in request error. Do not try to find transfer owner because there cannot
be one.

Make sure that the actual handling code path does not depend on the
configured log level because it is absolutely not desired for log level
to affect actual handling.

Signed-off-by: Tomasz Moń <tomasz.mon@nordicsemi.no>

tmon-nordic committed:
usb: device_next: Separate endpoint assignment from class data

Introduce usbd_class_iter for keeping endpoint assignment variables
and the single-linked list node. No functional changes right now, but
this paves the way for independent speed specific configurations.

Signed-off-by: Tomasz Moń <tomasz.mon@nordicsemi.no>

tmon-nordic committed:
usb: device_next: Fix uninitialized variable warning

Assign default value to prevent possible uninitialized variable warning.

Signed-off-by: Tomasz Moń <tomasz.mon@nordicsemi.no>

tmon-nordic committed:
usb: device_next: Allow class to return NULL descriptor set

Simply skip over the class if it returns NULL descriptor set for given
speed.

Signed-off-by: Tomasz Moń <tomasz.mon@nordicsemi.no>

tmon-nordic committed:
usb: device_next: Introduce speed specific configurations

USB High-Speed devices must be able to operate at both High-Speed and
Full-Speed. The USB specification allows the device to have different
configurations depending on connection speed. Modify the API to reflect
USB Specification requirements on what can (e.g. configurations) and
what cannot (e.g. VID, PID) be speed dependent.

While the class configurations for different speeds are completely
independent, the actual class instances are shared between operating
speeds (because only one speed can be active at a time). Classes are
free to provide different number of interfaces and/or endpoints for
different speeds. The endpoints are assigned for all operating speeds
during initialization.

Signed-off-by: Tomasz Moń <tomasz.mon@nordicsemi.no>

jfischer-no committed:
usb: device_next: temporarily remove struct usbd_class_data

With the latest change, there are two structures that actually contain
class instance related data. Merge them into usbd_class_dnode, because
there is already a handy macro that can be used in class instances.

Signed-off-by: Johann Fischer <johann.fischer@nordicsemi.no>

jfischer-no committed:
usb: device_next: rename usbd_class_node to usb_class_data

Since only usbd_class_node contains the class instance data, rename it
to usbd_class_data.

Signed-off-by: Johann Fischer <johann.fischer@nordicsemi.no>

jfischer-no committed:
usb: device_next: rename usbd_class_iter to usbd_class_node

For the last dozen commits, the role of usbd_class_node has actually
been taken over by usbd_class_iter. After cleaning up and merging
usbd_class_node and usbd_class_data, we can rename usbd_class_iter to
usbd_class_node to better reflect its role.

Signed-off-by: Johann Fischer <johann.fischer@nordicsemi.no>

jfischer-no committed:
usb: device_next: rework USBD shell to take speed into account

Rework USBD shell to take speed arguments into account and revise
command help text. Rename some of the commands to better reflect what
they actually do.
Since most of the init code changes, rewrite the command to initialize
the default configuration in the shell to register all available
classes, but not to enable USB device support in a single command.
This will allow earlier notification in the future, if supported by the
controller, before the USB device controller is enabled.

Signed-off-by: Johann Fischer <johann.fischer@nordicsemi.no>

jfischer-no committed:
usb: device_next: use helpers to get priv and device context

Some places have been overlooked, finally get rid of it. Use helpers to
get instance priv and device context.

Signed-off-by: Johann Fischer <johann.fischer@nordicsemi.no>

tmon-nordic commented on include/zephyr/usb/usbd.h:
```diff
@@ -195,68 +213,89 @@ struct usbd_cctx_vendor_req {
+};
+
+/**
+ * @cond INTERNAL_HIDDEN
```
It turns out that `/** @cond INTERNAL_HIDDEN */` must have a matching `/** @endcond */` as otherwise the build fails. Here the endcond should be right after `struct usbd_class_node`.

tmon-nordic commented on subsys/usb/device_next/class/usbd_msc.c:
```diff
@@ -724,30 +733,43 @@ static int msc_bot_request_handler(struct usbd_class_node *const node,
 {
-	struct msc_bot_ctx *ctx = node->data->priv;
+	struct msc_bot_ctx *ctx = usbd_class_get_private(c_data);
+	struct msc_bot_desc *desc = ctx->desc;
```
```suggestion
```

This fails build due to `usbd_msc.c:770:30: error: unused variable 'desc' [-Werror=unused-variable]`

tmon-nordic reviewed (changes requested):
Only few build failures remaining.

nordicjm commented on include/zephyr/usb/usbd.h:
```diff
@@ -141,6 +141,18 @@ struct usbd_ch9_data {
+ * @brief USB device speed
+ */
+enum usbd_speed {
+	/** Device suports or is connected to a full speed bus */
```
*supports

nordicjm commented on include/zephyr/usb/usbd.h:
```diff
@@ -141,6 +141,18 @@ struct usbd_ch9_data {
+enum usbd_speed {
+	/** Device suports or is connected to a full speed bus */
+	USBD_SPEED_FS,
+	/** Device suports or is connected to a high speed bus  */
+	USBD_SPEED_HS,
+	/** Device suports or is connected to a super speed bus */
```
as above

nordicjm commented on cmake/linker_script/common/common-ram.cmake:
```diff
@@ -121,7 +121,7 @@ endif()
   zephyr_iterable_section(NAME usb_cfg_data GROUP DATA_REGION ${XIP_ALIGN_WITH_INPUT} SUBALIGN 4)
   zephyr_iterable_section(NAME usbd_contex GROUP DATA_REGION ${XIP_ALIGN_WITH_INPUT} SUBALIGN 4)
-  zephyr_iterable_section(NAME usbd_class_node GROUP DATA_REGION ${XIP_ALIGN_WITH_INPUT} SUBALIGN 4)
+  zephyr_iterable_section(NAME usbd_class_iter GROUP DATA_REGION ${XIP_ALIGN_WITH_INPUT} SUBALIGN 4)
```
Missing changes in `include/zephyr/linker/common-ram.ld`

tmon-nordic replied:
What exactly do you mean? `usbd_class_node` never was in `include/zephyr/linker/common-ram.ld`, the changes are in `subsys/usb/device_next/usbd_data.ld`.

nordicjm replied:
Was not aware it was in a separate file

nordicjm commented on cmake/linker_script/common/common-ram.cmake:
```diff
@@ -121,7 +121,8 @@ endif()
   zephyr_iterable_section(NAME usbd_contex GROUP DATA_REGION ${XIP_ALIGN_WITH_INPUT} SUBALIGN 4)
-  zephyr_iterable_section(NAME usbd_class_iter GROUP DATA_REGION ${XIP_ALIGN_WITH_INPUT} SUBALIGN 4)
+  zephyr_iterable_section(NAME usbd_class_fs GROUP DATA_REGION ${XIP_ALIGN_WITH_INPUT} SUBALIGN 4)
+  zephyr_iterable_section(NAME usbd_class_hs GROUP DATA_REGION ${XIP_ALIGN_WITH_INPUT} SUBALIGN 4)
```
Missing changes in `include/zephyr/linker/common-ram.ld`

tmon-nordic replied:
This section is not in `include/zephyr/linker/common-ram.ld` but in `subsys/usb/device_next/usbd_data.ld`.

fabiobaltieri commented on include/zephyr/usb/usb_ch9.h:
```diff
@@ -321,6 +322,18 @@ struct usb_association_descriptor {
 #define USB_EP_TYPE_INTERRUPT		3U
 
+/** Calculate full speed interrupt endpoint bInterval from a value in microseconds */
+#define USB_FS_INT_EP_INTERVAL(us)	(CLAMP(((us) / 1000U), 1U, 255U))
```
you could get away with one less set of parenthesis here

fabiobaltieri commented on include/zephyr/usb/usbd.h:
```diff
@@ -141,6 +141,18 @@ struct usbd_ch9_data {
+ * @brief USB device speed
+ */
+enum usbd_speed {
+	/** Device supports or is connected to a full speed bus */
```
no low speed? (guess I'm missing some context here)

tmon-nordic replied:
I don't think there is any hardware in Zephyr that even supports Low-Speed as device, so it is not really possible to test if someone did implement Low-Speed at the stack side (the same can be said about the SS enum value here though).

fabiobaltieri replied:
ok fair enough, thanks for clarifying

fabiobaltieri commented:
@jfischer-no hey this needs a rebase
